server/src/main/java/org/elasticsearch/index/cache/bitset/BitsetFilterCache.java

```diff
@@ -92,10 +96,20 @@ public BitsetFilterCache(IndexSettings indexSettings, Listener listener) {
             throw new IllegalArgumentException("listener must not be null");
         }
         this.index = indexSettings.getIndex();
-        this.loadRandomAccessFiltersEagerly = indexSettings.getValue(INDEX_LOAD_RANDOM_ACCESS_FILTERS_EAGERLY_SETTING);
+        this.loadRandomAccessFiltersEagerly = shouldLoadRandomAccessFiltersEagerly(indexSettings);
         this.listener = listener;
     }
 
+    static boolean shouldLoadRandomAccessFiltersEagerly(IndexSettings settings) {
+        boolean loadFiltersEagerlySetting = settings.getValue(INDEX_LOAD_RANDOM_ACCESS_FILTERS_EAGERLY_SETTING);
+        boolean hasIndexRole = DiscoveryNode.hasRole(settings.getNodeSettings(), DiscoveryNodeRole.INDEX_ROLE);
+        boolean isStateless = DiscoveryNode.isStateless(settings.getNodeSettings());
+        if (isStateless && hasIndexRole) {
```

Comment:
Let's evaluate `hasIndexRole` after `isStateless` then I'd say :) Just so we don't pay for both checks during normal operation (not that it matters too much but still).